### PR TITLE
RPP Tensor Audio Support - Spectrogram

### PR DESCRIPTION
**Pull request by r-abishek**
- Adds Tensor Audio support on HOST for Spectrogram
- Adds relevant QA pass/fail tests and Performance tests

### CODE REVIEW

File: libs/third_party/ffts/libffts.a
Comment (rrawther) [inline libs/third_party/ffts/libffts.a:1]: why is this file added in the PR. Also do we need to refer to all these source files. Can we install this as a package?

File: src/modules/cpu/kernel/spectrogram.hpp [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+        length -= windowLength;
+    return ((length / windowStep) + 1);
+}
+
+// Compute reflect start idx to pad
+inline Rpp32s get_idx_reflect(Rpp32s loc, Rpp32s minLoc, Rpp32s maxLoc)
+{
+    if (maxLoc - minLoc < 2)
+        return maxLoc - 1;
+    for (;;)
```
Comment (rrawther) [inline src/modules/cpu/kernel/spectrogram.hpp:59]: why there is a infinite loop here? Not sure what this code is trying to do?

File: src/modules/cpu/kernel/spectrogram.hpp [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+                                  Rpp32s power,
+                                  Rpp32s windowLength,
+                                  Rpp32s windowStep,
+                                  RpptSpectrogramLayout layout,
+                                  rpp::Handle& handle)
+{
+    Rpp32s windowCenterOffset = 0;
+    bool vertical = (layout == RpptSpectrogramLayout::FT);
+    if (centerWindows)
+        windowCenterOffset = windowLength / 2;
```
Comment (rrawther) [inline src/modules/cpu/kernel/spectrogram.hpp:89]: move this to prev line

File: src/modules/cpu/kernel/spectrogram.hpp [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+    if (nfft == 0)
```
Comment (rrawther) [inline src/modules/cpu/kernel/spectrogram.hpp:90]: same here

**Review by rrawther — CHANGES_REQUESTED** [review]
please address review comments

File: src/modules/cpu/kernel/spectrogram.hpp (hunk 1 above)
Comment (snehaa8) [reply src/modules/cpu/kernel/spectrogram.hpp:59]: This loop will help converge on location between minLoc and maxLoc as per implementation in DALI.

File: libs/third_party/ffts/libffts.a
Comment (snehaa8) [reply libs/third_party/ffts/libffts.a:1]: DALI is using third party library for FFT(https://github.com/anthonix/ffts) and as we had to match outputs with DALI. We had included required files alone into our PR.

File: src/modules/cpu/kernel/spectrogram.hpp (hunk 2 above)
Comment (snehaa8) [reply src/modules/cpu/kernel/spectrogram.hpp:89]: Modified this

File: src/modules/cpu/kernel/spectrogram.hpp (hunk 3 above)
Comment (snehaa8) [reply src/modules/cpu/kernel/spectrogram.hpp:90]: Modified this

File: libs/third_party/ffts/libffts.a
Comment (r-abishek) [reply libs/third_party/ffts/libffts.a:1]: @rrawther We wanted RPP to be as independent as possible before, so this seems to be the best way considering its only one .a file needed. As we discussed before too, the whole library is actually much larger and has java code in it too. We don't need that, so the decision was to just plug in bare necessities needed here to get accuracy match on the training.

Comment (rrawther) [reply libs/third_party/ffts/libffts.a:1]: @r-abishek : we can't do this. You need to refer the code and make a new implementation from it  and include both .h and .cpp for it. Or import as a package which has to be available with ubuntu distribution. We can't ship RPP with .a library reference

File: src/modules/cpu/kernel/spectrogram.hpp [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+                    fftInBuf[off] = windowOutputTemp[i];
+                    fftInBuf[off + 1] = 0.0f;
+                }
+            }
+
+            ffts_execute(p, fftInBuf, fftOutBuf);
+            auto *complexFft = reinterpret_cast<std::complex<Rpp32f> *>(fftOutBuf);
+            for (int i = 0; i < numBins; i++)
+            {
+                if (vertical)
```
Comment (rrawther) [inline src/modules/cpu/kernel/spectrogram.hpp:213]: please remove the vertial/horizontal check outside the loop for better code

**Review by rrawther — CHANGES_REQUESTED** [review]
please address review comments

File: src/modules/cpu/kernel/spectrogram.hpp (hunk 4 above)
Comment (snehaa8) [reply src/modules/cpu/kernel/spectrogram.hpp:213]: Modified this, PR with fix will be pushed in soon.

**Review by rrawther — APPROVED** [review]

**Comment by LakshmiKumar23** [comment]
@r-abishek CI is failing on this. PLease have a look
```

+ cmake -DBACKEND=HIP -DCMAKE_BUILD_TYPE=Release ../..

-- The CXX compiler identification is Clang 14.0.6

-- Detecting CXX compiler ABI info

-- Detecting CXX compiler ABI info - done

-- Check for working CXX compiler: /usr/bin/clang++ - skipped

-- Detecting CXX compile features

-- Detecting CXX compile features - done

-- AMD ROCm Performance Primitives (RPP) Version -- 1.6.0

-- AMD RPP install path set to -- /opt/rocm

-- AMD RPP Backend set to -- HIP

-- AMD RPP Build Type -- Release

-- RPP Developer Options

--      -D BUILD_WITH_AMD_ADVANCE=ON [Turn ON/OFF Build for AMD advanced GPUs(default:ON)]

--      NOTE: For ROCm Version less than 6.0.0+ Use -D BUILD_WITH_AMD_ADVANCE=OFF

--      -D BACKEND=HIP [Select RPP Backend [options:CPU/OPENCL/HIP](default:HIP)]

--      -D CMAKE_BUILD_TYPE=Release [Select RPP build type [options:Debug/Release](default:Release)]

-- Found HALF: /usr/local/include  

-- Using HALF -- 

	Includes:/usr/local/include

-- Found OpenMP_CXX: -fopenmp=libomp (found version "5.0") 

-- Found OpenMP: TRUE (found version "5.0")  

-- Performing Test CMAKE_HAVE_LIBC_PTHREAD

-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success

-- Found Threads: TRUE  

-- Performing Test COMPILER_SUPPORTS_AVX

-- Performing Test COMPILER_SUPPORTS_AVX - Success

-- Performing Test HIP_CLANG_SUPPORTS_PARALLEL_JOBS

-- Performing Test HIP_CLANG_SUPPORTS_PARALLEL_JOBS - Failed

-- rpp -- Using HIP - Path:/opt/rocm	Version:6.2.40092	Compiler:clang

-- The C compiler identification is GNU 11.3.1

-- Detecting C compiler ABI info

-- Detecting C compiler ABI info - done

-- Check for working C compiler: /usr/bin/cc - skipped

-- Detecting C compile features

-- Detecting C compile features - done

-- modules -- Adding custom commands to rpp_kernels

-- HIP kernels added

-- AMD RPP modules -- Include Directories:/opt/rocm/include/;/var/jenkins_home/workspace/main_precheckin_rpp_PR-312/Q6BcBwnm3/rpp/src/include/hip/;/var/jenkins_home/workspace/main_precheckin_rpp_PR-312/Q6BcBwnm3/rpp/src/include/common/

-- rpp set to build with HIP backend

-- rpp set to build with ENABLE_SIMD_INTRINSICS

-- rpp -- Using Compiler - Path:/usr/bin/clang++	Version:14.0.6	Compiler:Clang

-- rpp -- CMAKE_CXX_FLAGS: -O3 -Ofast -DNDEBUG -fopenmp=libomp -fPIC -mavx2 -mf16c -mfma -std=gnu++17

-- rpp -- Link Libraries: stdc++fs;OpenMP::OpenMP_CXX;Threads::Threads;hip::host

-- rpp-test requires RPP. Install RPP before running CTests

-- AMD ROCm RPP RunTime Package --  rpp, rocm-hip-runtime-devel

-- AMD ROCm RPP Dev Package -- 

-- AMD ROCm RPP Test Package -- 

-- AMD ROCm RPP RunTime Package --  rocm-hip-runtime, half

-- AMD ROCm RPP Dev Package --  rpp, rocm-hip-runtime-dev

-- AMD ROCm RPP Test Package --  rpp-dev

-- Configuring done

-- Generating done

-- Build files have been written to: /var/jenkins_home/workspace/main_precheckin_rpp_PR-312/Q6BcBwnm3/rpp/build/release

++ nproc

+ make -j48

[  0%] Building CXX object addkernels/CMakeFiles/addkernels.dir/include_inliner.cpp.o

[  1%] Building CXX object addkernels/CMakeFiles/addkernels.dir/addkernels.cpp.o

[  2%] Linking CXX executable ../bin/addkernels

[  2%] Built target addkernels

[  3%] Inlining RPP kernels

[  3%] Building CXX object src/modules/CMakeFiles/modules.dir/binary_cache.cpp.o

[  4%] Building CXX object src/modules/CMakeFiles/modules.dir/logger.cpp.o

[  6%] Building CXX object src/modules/CMakeFiles/modules.dir/handle_api.cpp.o

[  6%] Building CXX object src/modules/CMakeFiles/modules.dir/md5.cpp.o

[  6%] Building CXX object src/modules/CMakeFiles/modules.dir/handlehost.cpp.o

[  7%] Building CXX object src/modules/CMakeFiles/modules.dir/rppi_advanced_augmentations.cpp.o

[  8%] Building CXX object src/modules/CMakeFiles/modules.dir/rppi_arithmetic_operations.cpp.o

[  9%] Building CXX object src/modules/CMakeFiles/modules.dir/rppi_color_model_conversions.cpp.o

[  9%] Building CXX object src/modules/CMakeFiles/modules.dir/rppi_filter_operations.cpp.o

[ 11%] Building CXX object src/modules/CMakeFiles/modules.dir/kernel_cache.cpp.o

[ 11%] Building CXX object src/modules/CMakeFiles/modules.dir/rppi_fused_functions.cpp.o

[ 12%] Building CXX object src/modules/CMakeFiles/modules.dir/rppi_geometry_transforms.cpp.o

[ 13%] Building CXX object src/modules/CMakeFiles/modules.dir/rppi_computer_vision.cpp.o

[ 14%] Building CXX object src/modules/CMakeFiles/modules.dir/rppi_image_augmentations.cpp.o

[ 15%] Building CXX object src/modules/CMakeFiles/modules.dir/rppi_logical_operations.cpp.o

[ 15%] Building CXX object src/modules/CMakeFiles/modules.dir/rppi_morphological_operations.cpp.o

[ 16%] Building CXX object src/modules/CMakeFiles/modules.dir/rppi_statistical_operations.cpp.o

[ 17%] Building CXX object src/modules/CMakeFiles/modules.dir/rppt_tensor_arithmetic_operations.cpp.o

[ 18%] Building CXX object src/modules/CMakeFiles/modules.dir/rppt_tensor_audio_augmentations.cpp.o

[ 19%] Building CXX object src/modules/CMakeFiles/modules.dir/rppt_tensor_color_augmentations.cpp.o

[ 19%] Building CXX object src/modules/CMakeFiles/modules.dir/rppt_tensor_data_exchange_operations.cpp.o

[ 20%] Building CXX object src/modules/CMakeFiles/modules.dir/rppt_tensor_filter_augmentations.cpp.o

[ 21%] Building CXX object src/modules/CMakeFiles/modules.dir/rppt_tensor_effects_augmentations.cpp.o

[ 22%] Building CXX object src/modules/CMakeFiles/modules.dir/rppt_tensor_geometric_augmentations.cpp.o

[ 23%] Building CXX object src/modules/CMakeFiles/modules.dir/rppt_tensor_logical_operations.cpp.o

[ 23%] Building CXX object src/modules/CMakeFiles/modules.dir/rppt_tensor_morphological_operations.cpp.o

[ 24%] Building CXX object src/modules/CMakeFiles/modules.dir/rppt_tensor_statistical_operations.cpp.o

[ 25%] Building CXX object src/modules/CMakeFiles/modules.dir/kernel.cpp.o

[ 25%] Building CXX object src/modules/CMakeFiles/modules.dir/hip/handlehip.cpp.o

[ 26%] Building CXX object src/modules/CMakeFiles/modules.dir/kernel_includes.cpp.o

[ 27%] Building CXX object src/modules/CMakeFiles/modules.dir/hip/hip_advanced_augmentation.cpp.o

[ 28%] Building CXX object src/modules/CMakeFiles/modules.dir/hip/hip_arithmetic_operations.cpp.o

[ 29%] Building CXX object src/modules/CMakeFiles/modules.dir/hip/hip_build_utils.cpp.o

[ 30%] Building CXX object src/modules/CMakeFiles/modules.dir/hip/hip_color_model_conversions.cpp.o

[ 30%] Building CXX object src/modules/CMakeFiles/modules.dir/hip/hip_computer_vision.cpp.o

[ 31%] Building CXX object src/modules/CMakeFiles/modules.dir/hip/hip_filter_functions.cpp.o

[ 32%] Building CXX object src/modules/CMakeFiles/modules.dir/hip/hip_fused_functions.cpp.o

[ 33%] Building CXX object src/modules/CMakeFiles/modules.dir/hip/hip_geometry_transforms.cpp.o

[ 34%] Building CXX object src/modules/CMakeFiles/modules.dir/hip/hip_image_augmentations.cpp.o

[ 34%] Building CXX object src/modules/CMakeFiles/modules.dir/hip/hip_logical_operations.cpp.o

[ 35%] Building CXX object src/modules/CMakeFiles/modules.dir/hip/hip_morphological_transforms.cpp.o

[ 36%] Building CXX object src/modules/CMakeFiles/modules.dir/hip/hip_statistical_operations.cpp.o

[ 37%] Building CXX object src/modules/CMakeFiles/modules.dir/hip/hip_utilities.cpp.o

[ 38%] Building CXX object src/modules/CMakeFiles/modules.dir/hip/hiperrors.cpp.o

[ 38%] Building CXX object src/modules/CMakeFiles/modules.dir/hip/hipoc_kernel.cpp.o

[ 39%] Building CXX object src/modules/CMakeFiles/modules.dir/hip/hipoc_program.cpp.o

[ 40%] Building CXX object src/modules/CMakeFiles/modules.dir/hip/tmp_dir.cpp.o

[ 41%] Building CXX object src/modules/CMakeFiles/modules.dir/hip/kernel/absolute_difference.cpp.o

/var/jenkins_home/workspace/main_precheckin_rpp_PR-312/Q6BcBwnm3/rpp/src/modules/rppt_tensor_audio_augmentations.cpp:203:1: error: function definition is not allowed here

  203 | {

      | ^

/var/jenkins_home/workspace/main_precheckin_rpp_PR-312/Q6BcBwnm3/rpp/src/modules/rppt_tensor_audio_augmentations.cpp:222:2: error: expected '}'

  222 | }

      |  ^

/var/jenkins_home/workspace/main_precheckin_rpp_PR-312/Q6BcBwnm3/rpp/src/modules/rppt_tensor_audio_augmentations.cpp:176:5: note: to match this '{'

  176 |     {

      |     ^

/var/jenkins_home/workspace/main_precheckin_rpp_PR-312/Q6BcBwnm3/rpp/src/modules/rppt_tensor_audio_augmentations.cpp:222:2: error: expected '}'

  222 | }

      |  ^

/var/jenkins_home/workspace/main_precheckin_rpp_PR-312/Q6BcBwnm3/rpp/src/modules/rppt_tensor_audio_augmentations.cpp:174:1: note: to match this '{'

  174 | {

      | ^

3 errors generated when compiling for gfx1030.

make[2]: *** [src/modules/CMakeFiles/modules.dir/build.make:406: src/modules/CMakeFiles/modules.dir/rppt_tensor_audio_augmentations.cpp.o] Error 1

make[2]: *** Waiting for unfinished jobs....

make[1]: *** [CMakeFiles/Makefile2:918: src/modules/CMakeFiles/modules.dir/all] Error 2

make: *** [Makefile:166: all] Error 2

+ sudo make install

[  2%] Built target addkernels

[  3%] Building CXX object src/modules/CMakeFiles/modules.dir/rppt_tensor_audio_augmentations.cpp.o

/var/jenkins_home/workspace/main_precheckin_rpp_PR-312/Q6BcBwnm3/rpp/src/modules/rppt_tensor_audio_augmentations.cpp:203:1: error: function definition is not allowed here

  203 | {

      | ^

/var/jenkins_home/workspace/main_precheckin_rpp_PR-312/Q6BcBwnm3/rpp/src/modules/rppt_tensor_audio_augmentations.cpp:222:2: error: expected '}'

  222 | }

      |  ^

/var/jenkins_home/workspace/main_precheckin_rpp_PR-312/Q6BcBwnm3/rpp/src/modules/rppt_tensor_audio_augmentations.cpp:176:5: note: to match this '{'

  176 |     {

      |     ^

/var/jenkins_home/workspace/main_precheckin_rpp_PR-312/Q6BcBwnm3/rpp/src/modules/rppt_tensor_audio_augmentations.cpp:222:2: error: expected '}'

  222 | }

      |  ^

/var/jenkins_home/workspace/main_precheckin_rpp_PR-312/Q6BcBwnm3/rpp/src/modules/rppt_tensor_audio_augmentations.cpp:174:1: note: to match this '{'

  174 | {

      | ^

3 errors generated when compiling for gfx1030.

make[2]: *** [src/modules/CMakeFiles/modules.dir/build.make:406: src/modules/CMakeFiles/modules.dir/rppt_tensor_audio_augmentations.cpp.o] Error 1

make[1]: *** [CMakeFiles/Makefile2:918: src/modules/CMakeFiles/modules.dir/all] Error 2

make: *** [Makefile:166: all] Error 2

+ sudo make package

[  2%] Built target addkernels

[  3%] Building CXX object src/modules/CMakeFiles/modules.dir/rppt_tensor_audio_augmentations.cpp.o

/var/jenkins_home/workspace/main_precheckin_rpp_PR-312/Q6BcBwnm3/rpp/src/modules/rppt_tensor_audio_augmentations.cpp:203:1: error: function definition is not allowed here

  203 | {

      | ^

/var/jenkins_home/workspace/main_precheckin_rpp_PR-312/Q6BcBwnm3/rpp/src/modules/rppt_tensor_audio_augmentations.cpp:222:2: error: expected '}'

  222 | }

      |  ^

/var/jenkins_home/workspace/main_precheckin_rpp_PR-312/Q6BcBwnm3/rpp/src/modules/rppt_tensor_audio_augmentations.cpp:176:5: note: to match this '{'

  176 |     {

      |     ^

/var/jenkins_home/workspace/main_precheckin_rpp_PR-312/Q6BcBwnm3/rpp/src/modules/rppt_tensor_audio_augmentations.cpp:222:2: error: expected '}'

  222 | }

      |  ^

/var/jenkins_home/workspace/main_precheckin_rpp_PR-312/Q6BcBwnm3/rpp/src/modules/rppt_tensor_audio_augmentations.cpp:174:1: note: to match this '{'

  174 | {

      | ^

3 errors generated when compiling for gfx1030.

make[2]: *** [src/modules/CMakeFiles/modules.dir/build.make:406: src/modules/CMakeFiles/modules.dir/rppt_tensor_audio_augmentations.cpp.o] Error 1

make[1]: *** [CMakeFiles/Makefile2:918: src/modules/CMakeFiles/modules.dir/all] Error 2

make: *** [Makefile:166: all] Error 2

+ ldd -v /opt/rocm/lib/librpp.so

ldd: /opt/rocm/lib/librpp.so: No such file or directory

script returned exit code 1
```

**Comment by r-abishek** [comment]
@kiritigowda @LakshmiKumar23 Merge conflicts resolved and CI passing

File: CMakeLists.txt [hunk 5]
```diff
@@ -300,6 +300,7 @@ message("-- ${White}${PROJECT_NAME} -- Link Libraries: ${LINK_LIBRARY_LIST}${Col
 target_link_libraries(${PROJECT_NAME} ${LINK_LIBRARY_LIST})
 set_target_properties(${PROJECT_NAME} PROPERTIES POSITION_INDEPENDENT_CODE ON)
 set_target_properties(${PROJECT_NAME} PROPERTIES LINKER_LANGUAGE CXX)
+target_link_libraries(${PROJECT_NAME} ${PROJECT_SOURCE_DIR}/libs/third_party/ffts/libffts.a)
```
Comment (rrawther) [inline CMakeLists.txt:303]: @r-abishek : I think we can't do this since source of this library is unknown and someone can possibly put a security hack here. Can you please replace this with
https://github.com/mreineck/pocketfft
It only uses header file and hence no source or so is required. You can just have the header file replicated in thirdparty with their license intact

**Review by rrawther — CHANGES_REQUESTED** [review]
added comment to change fft library

File: CMakeLists.txt (hunk 5 above)
Comment (r-abishek) [reply CMakeLists.txt:303]: @rrawther Yes that's true, and to that end, we'll check both these options:
- if a drop-in replacement with the above lib header works in terms of getting the same audio training convergence
- or if all necessary dependencies from the ".a" file could be extracted and added in RPP as raw C/C++ headers and built along with RPP.

**Review by rrawther — COMMENTED** [review]
@r-abishek : Here is another ubuntu fft lib package which can be used as an alternative
https://packages.ubuntu.com/source/focal/fftw
it can be installed as apt package

**Comment by rrawther** [comment]
@r-abishek : Any update on this PR. We need to merge all audio PRs soon in RPP. Since MIVisionX and rocAL depends on this, it is blocker for audio support which is due for ROCm 6.2

**Review by LakshmiKumar23 — APPROVED** [review]

**Comment by r-abishek** [comment]
@kiritigowda @rrawther @LakshmiKumar23 Removing WIP status on this. Can be merged if ffts copyright is good to use.

**Review by rrawther — APPROVED** [review]

**Comment by rrawther** [comment]
Kiriti, Can you merge this PR
